common/persistence/sql/sqlplugin/mysql/task.go

```diff
@@ -78,6 +78,12 @@ task_type = :task_type
 	rangeDeleteTaskQry = `DELETE FROM tasks ` +
 		`WHERE domain_id = ? AND task_list_name = ? AND task_type = ? AND task_id <= ? ` +
 		`ORDER BY domain_id,task_list_name,task_type,task_id LIMIT ?`
+
+	getOrphanTaskQry = `SELECT task_id, domain_id, task_list_name, task_type FROM tasks AS t ` +
+		`WHERE NOT EXISTS ( ` +
+		`	SELECT domain_id, name, task_type FROM task_lists AS tl ` +
+		`	WHERE t.domain_id=tl.domain_id and t.task_list_name=tl.name and t.task_type=tl.task_type ` +
```

Comment:
`shard_id` is the first column of the PK, it would be more efficient to pass in that value. 

The shardID is decided by domainId, tasklistName and nShards. So it should be easy to get: 
https://github.com/uber/cadence/blob/bfcb03fe53883ff4d574d9cc41e20dab202d4d52/common/persistence/sql/sqlTaskManager.go#L532


You may argue that will increase the complexity of the code.

Reply:
It's not clear to me how I could use this knowledge from code in order to update the query. It needs to scan over all tasks. I would argue the efficiency gained by using the index on task_lists is vastly overwhelmed by creating a need to make a separate client request per task so that the shard ID can be calculated in Go for each task row.

The cardinality of the task_lists table is relatively low; much less than the tasks table.

Reply:
You are right, I forgot that it's using `WHERE NOT EXISTS` so we can't map to a particular shardID.